### PR TITLE
Applying max_colwidth to the DataFrame index (#7856)

### LINKED ISSUE #7856: set_option('max_colwidth', N) not working on groupby output
Hi All,

Might be pilot error (or the incorrect expectations), but I was expecting the max_colwidth option to work on groupby terminal output. Replication here:
``` 
$ python
> import pandas as pd
> df = pd.DataFrame([{'a':'foo','b':'bar','c':'uncomfortably long line with lots of stuff','d':1},  \
                                   {'a':'foo','b':'bar','c':'stuff', 'd':1}])
> df
     a    b                                           c  d
0  foo  bar  uncomfortably long line with lots of stuff  1
1  foo  bar                                       stuff  1

> pd.set_option('max_colwidth', 20)
> df
     a    b                    c  d
0  foo  bar  uncomfortably lo...  1
1  foo  bar                stuff  1

> group_df = df.groupby(['a','b','c']).sum()
> group_df
                                                    d
a   b   c                                            
foo bar stuff                                       1
        uncomfortably long line with lots of stuff  1
>>> 
```

```
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.6.final.0
python-bits: 64
OS: Darwin
OS-release: 13.3.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.14.1
nose: None
Cython: 0.20.2
numpy: 1.8.1
scipy: None
statsmodels: None
IPython: 2.1.0
sphinx: 1.2.2
patsy: None
scikits.timeseries: None
dateutil: 2.2
pytz: 2014.4
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
rpy2: None
sqlalchemy: None
pymysql: None
psycopg2: None
>>> 
```

### PR DESCRIPTION
**Pull request by tiagoantao**
closes #7856 

reports a problem with the application of max_colwidth to groupby.

The real problem is that max_colwidth is not applied to the index. This patch solves that.

I am not applying self.justify to the formatting of the index (as that would break a test). But maybe it makes more sense to change the test?

### CODE REVIEW

**Comment by tiagoantao** [comment]
@jorisvandenbossche I now have time time to finalize this. I have merged with the current head. If there is still interest in dealing with this issue just say. I could imagine that there is still a few things to do (like adding an entry to whatsnew)...

**Comment by jorisvandenbossche** [comment]
@tiagoantao Certainly interest!

Remark on the test: I don't think it is testing the correct thing (but you can also leave it in as it is now for testing the width of the columns), as it is testing the width of the columns which is already working. You can eg set some columns as the index, and then test again. That is what now is failing:

```
In [25]: pd.set_option('max_colwidth', 20)

In [26]: df
Out[26]: 
     a    b                    c  d
0  foo  bar  uncomfortably lo...  1
1  foo  bar                stuff  1

In [27]: df.set_index(['a', 'b', 'c'])
Out[27]: 
                                                    d
a   b   c                                            
foo bar uncomfortably long line with lots of stuff  1
        stuff                                       1
```

File: pandas/tests/test_frame.py [hunk 1]
```diff
@@ -4795,6 +4795,22 @@ def test_repr_unicode(self):
         result = repr(df)
         self.assertEqual(result.split('\n')[0].rstrip(), ex_top)
 
+    def test_str_max_colwidth(self):
+        curr_max_colwidth = pd.get_option('max_colwidth')
+        # As we change a global option, we save it
```
Comment (jreback) [inline pandas/tests/test_frame.py]: pls add the issue number as a comment here

**Comment by jreback** [comment]
pls add a release note that explains this. if you can do this asap i can put in 0.16.0

**Comment by jorisvandenbossche** [comment]
And the test needs an addition, see my comment above.

**Comment by tiagoantao** [comment]
I will do this over the weekend.

On 6 March 2015 at 09:21, Joris Van den Bossche <notifications@github.com>
wrote:

> And the test needs an addition, see my comment above.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/8954#issuecomment-77529226>.
>



-- 
When the senses are shaken, and the soul is driven to madness, Who can
stand? - William Blake

**Comment by tiagoantao** [comment]
I have tried to sort this out. My sincere apologies for the delay. I will be monitoring this during the next day or so. I will try to react as fast as possible if you find anything in need of intervention.

File: pandas/tests/test_frame.py [hunk 2]
```diff
@@ -4815,6 +4815,24 @@ def test_repr_unicode(self):
...
+    def test_str_max_colwidth(self):
+        # GH 7856
+        curr_max_colwidth = pd.get_option('max_colwidth')
+        # As we change a global option, we save it
+        df = pd.DataFrame([{'a': 'foo', 'b': 'bar',
+                            'c': 'uncomfortably long line with lots of stuff',
+                            'd': 1},
+                           {'a': 'foo', 'b': 'bar', 'c': 'stuff', 'd': 1}])
+        df.set_index(['a', 'b', 'c'])
+        assert(str(df) == '     a    b                                           c  d\n'
```
Comment (jreback) [inline pandas/tests/test_frame.py]: use ``self.assertTrue(.....)`` instead of bare assert

File: pandas/tests/test_frame.py [hunk 3]
```diff
@@ -4815,6 +4815,24 @@ def test_repr_unicode(self):
         result = repr(df)
         self.assertEqual(result.split('\n')[0].rstrip(), ex_top)
 
+    def test_str_max_colwidth(self):
+        # GH 7856
+        curr_max_colwidth = pd.get_option('max_colwidth')
```
Comment (jreback) [inline pandas/tests/test_frame.py]: these tests should all be in test_format.py instead of here

File: pandas/tests/test_frame.py [hunk 4]
```diff
@@ -4815,6 +4815,24 @@ def test_repr_unicode(self):
         result = repr(df)
         self.assertEqual(result.split('\n')[0].rstrip(), ex_top)
 
+    def test_str_max_colwidth(self):
+        # GH 7856
```
Comment (jreback) [inline pandas/tests/test_frame.py]: use ``with pd.option_context('max_colwidth',.....):`` so that it will only affect the code you want it to

**Comment by tiagoantao** [comment]
Did all the corrections. I just hope the way I committed (i.e. in terms of logs and squash) is acceptable. My git-fu is a bit substandard.

I apologize for all these mistakes. I hope in the future, with this experience, I am able to do a few contributions will less noise.

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/e4cb0f8a6cbb5f0c89b24783baa44326e4b2cccb

thanks!

**Pull request closed without merging**